backend/cn/lib/eval.ml

```diff
@@ -0,0 +1,607 @@
+    Error ("Mismatched signs at " ^ Locations.simple_location here)
+  else if sz1 <> sz2 then
+    Error ("Mismatched sizes at " ^ Locations.simple_location here)
+  else
+    Ok ()
+
+
+let eval_num_binop
+  (eval : IT.t -> (IT.t, string) result)
+  (bt : BT.t)
+  (here : Locations.t)
+  (f : Z.t -> Z.t -> Z.t)
+  (it1 : IT.t)
+  (it2 : IT.t)
+  (loc : string)
+  : (IT.t, string) result
+  =
+  let ( let@ ) = Result.bind in
+  let return = Result.ok in
+  let@ it1 = eval it1 in
+  let@ it2 = eval it2 in
+  match (it1, it2) with
+  | IT (Const (Z n1), _, _), IT (Const (Z n2), _, _) ->
+    return @@ IT.num_lit_ (f n1 n2) bt here
+  | IT (Const (Bits ((sgn1, sz1), n1)), _, _), IT (Const (Bits ((sgn2, sz2), n2)), _, _)
+    ->
+    let@ () = check_bits_bt (sgn1, sz1) (sgn2, sz2) here in
+    return @@ IT.num_lit_ (BT.normalise_to_range_bt bt (f n1 n2)) bt here
+  | _, IT (Const (Z _), _, _) ->
+    Error ("Not constant integer `" ^ Pp.plain (IT.pp it1) ^ "` (" ^ loc ^ ")")
+  | _, IT (Const (Bits _), _, _) ->
+    Error ("Not constant bitvector `" ^ Pp.plain (IT.pp it1) ^ "` (" ^ loc ^ ")")
+  | IT (Const (Z _), _, _), _ ->
+    Error ("Not constant integer `" ^ Pp.plain (IT.pp it2) ^ "` (" ^ loc ^ ")")
+  | IT (Const (Bits _), _, _), _ ->
+    Error ("Not constant bitvector `" ^ Pp.plain (IT.pp it2) ^ "` (" ^ loc ^ ")")
+  | _, _ ->
+    Error
+      ("Not constant integer or bitvector `"
+       ^ Pp.plain (IT.pp it1)
+       ^ "` and `"
+       ^ Pp.plain (IT.pp it2)
+       ^ "` ("
+       ^ loc
+       ^ ")")
+
+
+let eval_term_generic
+  (eval_aux : IT.t -> (IT.t, string) result)
+  (prog5 : unit Mucore.mu_file)
+  (it : IT.t)
+  : (IT.t, string) result
+  =
+  let ( let@ ) = Result.bind in
+  let return = Result.ok in
+  let open IT in
+  let (IT (t_, bt, here)) = it in
+  let eval_num_binop = eval_num_binop eval_aux bt here in
+  match t_ with
+  | Const _ | Nil _ | MapDef _ -> return it
+  | Sym x -> Error (Sym.pp_string x ^ " is free")
+  (* Unary ops *)
+  | Unop (Not, it') ->
+    let@ it' = eval_aux it' in
+    (match it' with
+     | IT (Const (Bool b), _, _) -> return @@ bool_ (not b) here
+     | _ -> Error ("Not constant boolean (" ^ __LOC__ ^ ")"))
+  | Unop (Negate, it') ->
+    let@ it' = eval_aux it' in
+    (match it' with
+     | IT (Const (Z n), _, _) -> return @@ num_lit_ (Z.neg n) bt here
+     | IT (Const (Bits ((Signed, _), n)), _, _) ->
+       return @@ num_lit_ (BT.normalise_to_range_bt bt (Z.neg n)) bt here
+     | IT (Const (Bits ((Unsigned, _), _)), _, _) ->
+       Error
+         ("Can't negate unsigned integer at "
+          ^ Locations.simple_location here
+          ^ " ("
+          ^ __LOC__
+          ^ ")")
+     | _ -> Error ("Not constant integer or bitvector (" ^ __LOC__ ^ ")"))
+  | Unop (BW_CLZ_NoSMT, it') ->
+    let@ it' = eval_aux it' in
+    (match it' with
+     | IT (Const (Bits ((sgn, bits), n)), bt, _) ->
+       let open Int64 in
+       let reverse_bits (n : Z.t) : Z.t =
```

Comment:
I'm not sure this does the right thing, at least for bitvector sizes larger than 64?